File: src/ffi.rs
use cxx::{type_id, ExternType};

use crate::bindings;

pub use self::ffi::*;

pub struct RequestCompleteContext {
    pub handler: Box<dyn Fn(&Request) + Send + Sync + 'static>,
}

unsafe impl ExternType for bindings::libcamera_StreamRole {
    type Id = type_id!("libcamera::StreamRole");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_Request_Status {
    type Id = type_id!("libcamera::RequestStatus");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_Request_ReuseFlag {
    type Id = type_id!("libcamera::RequestReuseFlag");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_FrameMetadata_Status {
    type Id = type_id!("libcamera::FrameStatus");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_FrameMetadata_Plane {
    type Id = type_id!("libcamera::FramePlaneMetadata");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_CameraConfiguration_Status {
    type Id = type_id!("libcamera::CameraConfigurationStatus");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_PixelFormat {
    type Id = type_id!("libcamera::PixelFormat");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_Size {
    type Id = type_id!("libcamera::Size");
    type Kind = cxx::kind::Trivial;
}

unsafe impl ExternType for bindings::libcamera_SizeRange {
    type Id = type_id!("libcamera::SizeRange");
    type Kind = cxx::kind::Trivial;
}

// libcamera_CameraConfiguration_Status

#[cxx::bridge]
mod ffi {
    /// A mirror of libcamera::FrameBuffer::Plane
    #[derive(Debug, Clone, Copy)]
    struct FrameBufferPlane {
        fd: u32,
        offset: u32,
        length: u32,
    }

    #[derive(Debug, Clone)]
    struct FrameMetadata {
        status: FrameStatus,
        sequence: u32,
        timestamp: u64,
        planes: Vec<FramePlaneMetadataWrap>,
    }

    // Wrapper to work around https://github.com/dtolnay/cxx/issues/741
    struct CameraPtr {
        camera: SharedPtr<Camera>,
    }

    struct FrameBufferPtr {
        buffer: *mut FrameBuffer,
    }

    #[derive(Debug, Clone)]
    struct FramePlaneMetadataWrap {
        inner: FramePlaneMetadata,
    }

    struct StreamPtr {
        stream: *mut Stream,
    }

    struct PixelFormatWrap {
        value: PixelFormat,
    }

    struct SizeWrap {
        value: Size,
    }

    extern "Rust" {
        type RequestCompleteContext;
    }

    #[namespace = "libcamera"]
    unsafe extern "C++" {
        include!("libcamera/src/ffi.h");

        // Keep in sync with ExternType definitions above.
        type StreamRole = crate::bindings::libcamera_StreamRole;
        type RequestStatus = crate::bindings::libcamera_Request_Status;
        type RequestReuseFlag = crate::bindings::libcamera_Request_ReuseFlag;
        type FrameStatus = crate::bindings::libcamera_FrameMetadata_Status;
        type FramePlaneMetadata = crate::bindings::libcamera_FrameMetadata_Plane;
        type CameraConfigurationStatus = crate::bindings::libcamera_CameraConfiguration_Status;
        type PixelFormat = crate::bindings::libcamera_PixelFormat;
        type Size = crate::bindings::libcamera_Size;
        type SizeRange = crate::bindings::libcamera_SizeRange;

        //////////////////////////////////////

        type CameraManager;

        fn new_camera_manager() -> UniquePtr<CameraManager>;

        fn start(self: Pin<&mut CameraManager>) -> i32;

        fn stop(self: Pin<&mut CameraManager>);

        fn list_cameras(camera_manager: &CameraManager) -> Vec<CameraPtr>;

        //////////////////////////////////////

        type Camera;

        fn id(self: &Camera) -> &CxxString;

        /// Thread safe
        fn acquire(self: Pin<&mut Camera>) -> i32;

        fn release(self: Pin<&mut Camera>) -> i32;

        // Thread safe
        fn generate_camera_configuration(
            camera: Pin<&mut Camera>,
            stream_roles: &[StreamRole],
        ) -> UniquePtr<CameraConfiguration>;

        unsafe fn start(self: Pin<&mut Camera>, control_list: *const ControlList) -> i32;

        unsafe fn configure(self: Pin<&mut Camera>, config: *mut CameraConfiguration) -> i32;

        /// Thread safe
        fn createRequest(self: Pin<&mut Camera>, cookie: u64) -> UniquePtr<Request>;

        /// Thread safe
        unsafe fn queueRequest(self: Pin<&mut Camera>, request: *mut Request) -> i32;

        /// Thread safe
        type RequestCompleteSlot;
        fn camera_connect_request_completed(
            camera: Pin<&mut Camera>,
            handler: fn(&RequestCompleteContext, &Request),
            context: Box<RequestCompleteContext>,
        ) -> UniquePtr<RequestCompleteSlot>;

        fn camera_streams(camera: &Camera) -> Vec<StreamPtr>;

        unsafe fn camera_contains_stream(camera: &Camera, stream: *mut Stream) -> bool;

        //////////////////////////////////////

        type Request;

        fn sequence(self: &Request) -> u32;
        fn cookie(self: &Request) -> u64;

        // May return a negative error.
        unsafe fn addBuffer(
            self: Pin<&mut Request>,
            stream: *const Stream,
            buffer: *mut FrameBuffer,
            fence: UniquePtr<Fence>,
        ) -> i32;

        fn status(self: &Request) -> RequestStatus;

        fn reuse(self: Pin<&mut Request>, flags: RequestReuseFlag);

        /*

        ControlList &controls() { return *controls_; }
        ControlList &metadata() { return *metadata_; }
        const BufferMap &buffers() const { return bufferMap_; }
        int addBuffer(const Stream *stream, FrameBuffer *buffer,
                      std::unique_ptr<Fence> fence = nullptr);
        FrameBuffer *findBuffer(const Stream *stream) const;

        bool hasPendingBuffers() const;

        std::string toString() const;

        */

        //////////////////////////////////////

        type Fence;

        //////////////////////////////////////

        type CameraConfiguration;

        fn at(self: &CameraConfiguration, index: u32) -> &StreamConfiguration;

        #[rust_name = "at_mut"]
        fn at(self: Pin<&mut CameraConfiguration>, index: u32) -> Pin<&mut StreamConfiguration>;

        fn size(self: &CameraConfiguration) -> usize;

        fn validate(self: Pin<&mut CameraConfiguration>) -> CameraConfigurationStatus;

        //////////////////////////////////////

        type StreamConfiguration;

        fn stream_config_pixel_format(config: &StreamConfiguration) -> PixelFormat;
        fn stream_config_set_pixel_format(
            config: Pin<&mut StreamConfiguration>,
            value: PixelFormat,
        );

        fn stream_config_size(config: &StreamConfiguration) -> Size;
        fn stream_config_set_size(config: Pin<&mut StreamConfiguration>, value: Size);

        fn stream_config_stride(config: &StreamConfiguration) -> u32;
        fn stream_config_set_stride(config: Pin<&mut StreamConfiguration>, value: u32);

        fn stream_config_frame_size(config: &StreamConfiguration) -> u32;
        fn stream_config_set_frame_size(config: Pin<&mut StreamConfiguration>, value: u32);

        fn stream_config_buffer_count(config: &StreamConfiguration) -> u32;
        fn stream_config_set_buffer_count(config: Pin<&mut StreamConfiguration>, value: u32);

        fn stream_config_to_string(config: &StreamConfiguration) -> String;

        fn stream(self: &StreamConfiguration) -> *mut Stream;

        fn formats(self: &StreamConfiguration) -> &StreamFormats;

        //////////////////////////////////////

        type Stream;

        //////////////////////////////////////

        fn pixel_format_to_string(format: &PixelFormat) -> String;

        //////////////////////////////////////

        type StreamFormats;

        fn stream_formats_pixelformats(stream_formats: &StreamFormats) -> Vec<PixelFormatWrap>;

        fn stream_formats_sizes(
            stream_formats: &StreamFormats,
            pixelformat: &PixelFormat,
        ) -> Vec<SizeWrap>;

        fn range(self: &StreamFormats, pixelformat: &PixelFormat) -> SizeRange;

        //////////////////////////////////////

        type ControlList;

        // ControlId { id(), name(), type() }

        //////////////////////////////////////

        type FrameBuffer;

        fn frame_buffer_planes(buffer: &FrameBuffer) -> Vec<FrameBufferPlane>;

        fn frame_buffer_metadata(buffer: &FrameBuffer) -> FrameMetadata;

        //////////////////////////////////////

        type FrameBufferAllocator;

        fn new_frame_buffer_allocator(camera: SharedPtr<Camera>)
            -> UniquePtr<FrameBufferAllocator>;

        unsafe fn allocate(self: Pin<&mut FrameBufferAllocator>, stream: *mut Stream) -> i32;

        unsafe fn free(self: Pin<&mut FrameBufferAllocator>, stream: *mut Stream) -> i32;

        unsafe fn get_allocated_frame_buffers(
            allocator: &FrameBufferAllocator,
            stream: *mut Stream,
        ) -> Vec<FrameBufferPtr>;

        //////////////////////////////////////

    }
}

File: src/ffi.cc
#include "ffi.h"

namespace libcamera {

std::unique_ptr<CameraManager> new_camera_manager() {
  return std::make_unique<CameraManager>();
}

rust::Vec<CameraPtr> list_cameras(const CameraManager &camera_manager) {
  auto cameras = camera_manager.cameras();

  rust::Vec<CameraPtr> out;
  out.reserve(cameras.size());
  for (const auto &camera : cameras) {
    out.push_back(CameraPtr{.camera = camera});
  }

  return out;
}

std::unique_ptr<CameraConfiguration> generate_camera_configuration(
    Camera &camera, rust::Slice<const StreamRole> stream_roles) {
  StreamRoles roles;
  for (const auto &role : stream_roles) {
    roles.push_back(role);
  }

  auto config = camera.generateConfiguration(roles);

  return config;
}

std::unique_ptr<FrameBufferAllocator> new_frame_buffer_allocator(
    std::shared_ptr<Camera> camera) {
  return std::make_unique<FrameBufferAllocator>(camera);
}

rust::String stream_config_to_string(const StreamConfiguration &config) {
  return rust::String(config.toString());
}

PixelFormat stream_config_pixel_format(const StreamConfiguration &config) {
  return config.pixelFormat;
}
void stream_config_set_pixel_format(StreamConfiguration &config,
                                    PixelFormat value) {
  config.pixelFormat = value;
}

Size stream_config_size(const StreamConfiguration &config) {
  return config.size;
}
void stream_config_set_size(StreamConfiguration &config, Size value) {
  config.size = value;
}

unsigned int stream_config_stride(const StreamConfiguration &config) {
  return config.stride;
}
void stream_config_set_stride(StreamConfiguration &config, unsigned int value) {
  config.stride = value;
}

unsigned int stream_config_frame_size(const StreamConfiguration &config) {
  return config.frameSize;
}
void stream_config_set_frame_size(StreamConfiguration &config,
                                  unsigned int value) {
  config.frameSize = value;
}

unsigned int stream_config_buffer_count(const StreamConfiguration &config) {
  return config.bufferCount;
}
void stream_config_set_buffer_count(StreamConfiguration &config,
                                    unsigned int value) {
  config.bufferCount = value;
}

rust::Vec<FrameBufferPtr> get_allocated_frame_buffers(
    const FrameBufferAllocator &allocator, Stream *stream) {
  const auto &buffers = allocator.buffers(stream);

  rust::Vec<FrameBufferPtr> out;
  out.reserve(buffers.size());

  for (const auto &buffer : buffers) {
    out.push_back(FrameBufferPtr{.buffer = buffer.get()});
  }

  return out;
}

rust::Vec<::FrameBufferPlane> frame_buffer_planes(const FrameBuffer &buffer) {
  rust::Vec<::FrameBufferPlane> out;

  for (const auto &plane : buffer.planes()) {
    out.push_back(
        ::FrameBufferPlane{.fd = static_cast<unsigned int>(plane.fd.get()),
                           .offset = plane.offset,
                           .length = plane.length});
  }

  return out;
}

::FrameMetadata frame_buffer_metadata(const FrameBuffer &buffer) {
  const auto &meta = buffer.metadata();

  ::FrameMetadata out{.status = meta.status,
                      .sequence = meta.sequence,
                      .timestamp = meta.timestamp,
                      .planes = rust::Vec<::FramePlaneMetadataWrap>()};

  for (const auto &plane : meta.planes()) {
    out.planes.push_back({.inner = plane});
  }

  return out;
}

rust::String pixel_format_to_string(const PixelFormat &format) {
  return rust::String(format.toString());
}

rust::Vec<PixelFormatWrap> stream_formats_pixelformats(
    const StreamFormats &stream_formats) {
  auto value = stream_formats.pixelformats();
  rust::Vec<PixelFormatWrap> out;
  for (const auto &v : value) {
    out.push_back({.value = v});
  }
  return out;
}
rust::Vec<SizeWrap> stream_formats_sizes(const StreamFormats &stream_formats,
                                         const PixelFormat &pixelformat) {
  auto value = stream_formats.sizes(pixelformat);
  rust::Vec<SizeWrap> out;
  for (const auto &v : value) {
    out.push_back({.value = v});
  }
  return out;
}

std::unique_ptr<RequestCompleteSlot> camera_connect_request_completed(
    Camera &camera,
    rust::Fn<void(const RequestCompleteContext &, const Request &)> handler,
    rust::Box<::RequestCompleteContext> context) {
  auto slot = std::make_unique<RequestCompleteSlot>(
      &camera.requestCompleted, std::move(handler), std::move(context));

  return slot;
}

rust::Vec<::StreamPtr> camera_streams(const Camera &camera) {
  rust::Vec<::StreamPtr> out;
  for (auto stream : camera.streams()) {
    out.push_back({.stream = stream});
  }
  return out;
}

bool camera_contains_stream(const Camera &camera, Stream *stream) {
  return camera.streams().contains(stream);
}

rust::String request_to_string(const Request &request) {
  return rust::String(request.toString());
}

rust::Vec<::ControlInfoMapEntry> control_info_map_entries(
    const ControlInfoMap &map) {
  rust::Vec<::ControlInfoMapEntry> out;
  for (const auto &[key, value] : map) {
    out.push_back(::ControlInfoMapEntry{.key = *key, .value = value});
  }

  return out;
}

rust::String control_value_get_string(const ControlValue &value) {
  return rust::String(value.get<std::string>());
}

void control_value_set_string(ControlValue &value, const rust::String &s) {
  value.set<std::string>(std::string(s));
}

rust::Vec<rust::String> control_value_get_string_array(
    const ControlValue &value) {
  rust::Vec<rust::String> out;
  for (const auto &v : value.get<Span<const std::string>>()) {
    out.push_back(v);
  }

  return out;
}

rust::Vec<::ControlListEntry> control_list_entries(const ControlList &list) {
  rust::Vec<::ControlListEntry> out;
  for (const auto &[key, value] : list) {
    out.push_back(::ControlListEntry{.key = key, .value = value});
  }
  return out;
}

rust::String control_value_to_string(const ControlValue &value) {
  return rust::String(value.toString());
}

rust::String control_info_to_string(const ControlInfo &info) {
  return rust::String(info.toString());
}

}  // namespace libcamera
